pkg/version/version.go

```diff
@@ -0,0 +1,307 @@
+	"github.com/replicatedhq/kotsadm/pkg/downstream"
+	"github.com/replicatedhq/kotsadm/pkg/gitops"
+	"github.com/replicatedhq/kotsadm/pkg/kotsutil"
+	"github.com/replicatedhq/kotsadm/pkg/persistence"
+	registrytypes "github.com/replicatedhq/kotsadm/pkg/registry/types"
+	"github.com/replicatedhq/kotsadm/pkg/render"
+	"github.com/replicatedhq/kotsadm/pkg/version/types"
+)
+
+// CreateFirstVersion works much likst CreateVersion except that it assumes version 0
+// and never attempts to calculate a diff, or look at previous versions
+func CreateFirstVersion(appID string, filesInDir string, source string) (int64, error) {
+	return createVersion(appID, filesInDir, source, nil)
+}
+
+// CreateVersion creates a new version of the app in the database, but the caller
+// is responsible for uploading the archive to s3
+func CreateVersion(appID string, filesInDir string, source string, currentSequence int64) (int64, error) {
+	return createVersion(appID, filesInDir, source, &currentSequence)
+}
+
+// this is the common, internal function to create an app version, used in both
+// new and updates to apps
+func createVersion(appID string, filesInDir string, source string, currentSequence *int64) (int64, error) {
+	kotsKinds, err := kotsutil.LoadKotsKindsFromPath(filesInDir)
+	if err != nil {
+		return int64(0), errors.Wrap(err, "failed to read kots kinds")
+	}
+
+	supportBundleSpec, err := kotsKinds.Marshal("troubleshoot.replicated.com", "v1beta1", "Collector")
+	if err != nil {
+		return int64(0), errors.Wrap(err, "failed to marshal support bundle spec")
+	}
+	analyzersSpec, err := kotsKinds.Marshal("troubleshoot.replicated.com", "v1beta1", "Analyzer")
+	if err != nil {
+		return int64(0), errors.Wrap(err, "failed to marshal analyzer spec")
+	}
+	preflightSpec, err := kotsKinds.Marshal("troubleshoot.replicated.com", "v1beta1", "Preflight")
+	if err != nil {
+		return int64(0), errors.Wrap(err, "failed to marshal preflight spec")
+	}
+
+	appSpec, err := kotsKinds.Marshal("app.k8s.io", "v1beta1", "Application")
+	if err != nil {
+		return int64(0), errors.Wrap(err, "failed to marshal app spec")
+	}
+	kotsAppSpec, err := kotsKinds.Marshal("kots.io", "v1beta1", "Application")
+	if err != nil {
+		return int64(0), errors.Wrap(err, "failed to marshal kots app spec")
+	}
+	backupSpec, err := kotsKinds.Marshal("velero.io", "v1", "Backup")
+	if err != nil {
+		return int64(0), errors.Wrap(err, "failed to marshal backup spec")
+	}
+
+	licenseSpec, err := kotsKinds.Marshal("kots.io", "v1beta1", "License")
+	if err != nil {
+		return int64(0), errors.Wrap(err, "failed to marshal license spec")
+	}
+	configSpec, err := kotsKinds.Marshal("kots.io", "v1beta1", "Config")
+	if err != nil {
+		return int64(0), errors.Wrap(err, "failed to marshal config spec")
+	}
+	configValuesSpec, err := kotsKinds.Marshal("kots.io", "v1beta1", "ConfigValues")
+	if err != nil {
+		return int64(0), errors.Wrap(err, "failed to marshal configvalues spec")
+	}
+
+	// TODO: solve
```

Comment:
this TODO is very vague

Reply:
ha! i fixed the code but forgot to remove the todo!